From: Stefan Bader <stefan.bader@canonical.com>
Date: Wed, 24 Jul 2013 15:26:45 +0200
Subject: Use alternate method to check for running Xen legacy toolstack

Upstream runs "/usr/sbin/xend status" but we have (at least for some
time) shipped xend only in the private binary path (/usr/lib/xen-*/bin).
And that path depends on the Xen version. Instead modify the check to
call "/usr/lib/xen-common/xen-toolstack" which does not change between
Xen versions and returns the path to xm or xl. Then decide on the base-
name (xm) whether the legacy toolstack is active.

This unlikely will be accepted by upstream libvirt as obviously they
expect a different packaging.

For Ubuntu I also had to add "/usr/lib/xen-common/xen-toolstack" to
the allowed callouts of apparmor.

Signed-off-by: Stefan Bader <stefan.bader@canonical.com>
---
 src/libxl/libxl_driver.c | 23 ++++++++++++++++++-----
 src/xen/xen_driver.c     | 17 ++++++++++++++---
 2 files changed, 32 insertions(+), 8 deletions(-)

diff --git a/src/libxl/libxl_driver.c b/src/libxl/libxl_driver.c
index a79efcb..4a11e81 100644
--- a/src/libxl/libxl_driver.c
+++ b/src/libxl/libxl_driver.c
@@ -941,6 +941,7 @@ libxlDriverShouldLoad(bool privileged)
 {
     bool ret = false;
     virCommandPtr cmd;
+    char *output;
     int status;
 
     /* Don't load if non-root */
@@ -958,14 +959,26 @@ libxlDriverShouldLoad(bool privileged)
     }
 
     /* Don't load if legacy xen toolstack (xend) is in use */
-    cmd = virCommandNewArgList("/usr/sbin/xend", "status", NULL);
+    cmd = virCommandNewArgList("/usr/lib/xen-common/bin/xen-toolstack", NULL);
+    virCommandSetOutputBuffer(cmd, &output);
     if (virCommandRun(cmd, &status) == 0 && status == 0) {
-        VIR_INFO("Legacy xen tool stack seems to be in use, disabling "
-                  "libxenlight driver.");
-    } else {
-        ret = true;
+        int i, j;
+
+        for (i = 0, j = 0; output[i] != '\0'; i++)
+            if (output[i] == '/')
+                j = i + 1;
+
+        if (output[j] == 'x' && output[j+1] == 'm') {
+            VIR_INFO("Legacy xen tool stack seems to be in use, disabling "
+                     "libxenlight driver.");
+            VIR_FREE(output);
+            virCommandFree(cmd);
+            return ret;
+        }
     }
+    VIR_FREE(output);
     virCommandFree(cmd);
+    ret = true;
 
     return ret;
 }
diff --git a/src/xen/xen_driver.c b/src/xen/xen_driver.c
index 7506e8c..85a668f 100644
--- a/src/xen/xen_driver.c
+++ b/src/xen/xen_driver.c
@@ -312,12 +312,23 @@ static int
 xenUnifiedXendProbe(void)
 {
     virCommandPtr cmd;
+    char *output;
     int status;
     int ret = 0;
 
-    cmd = virCommandNewArgList("/usr/sbin/xend", "status", NULL);
-    if (virCommandRun(cmd, &status) == 0 && status == 0)
-        ret = 1;
+    cmd = virCommandNewArgList("/usr/lib/xen-common/bin/xen-toolstack", NULL);
+    virCommandSetOutputBuffer(cmd, &output);
+    if (virCommandRun(cmd, &status) == 0 && status == 0) {
+        int i, j;
+
+        for (i = 0, j = 0; output[i] != '\0'; i++)
+            if (output[i] == '/')
+                j = i + 1;
+
+        if (output[j] == 'x' && output[j+1] == 'm')
+            ret = 1;
+    }
+    VIR_FREE(output);
     virCommandFree(cmd);
 
     return ret;
